sdks/go/pkg/beam/io/bigtableio/bigtable.go

```diff
@@ -0,0 +1,278 @@
+	// Client is the bigtable.Client
+	Client *bigtable.Client `json:"client"`
+	// TableName is the qualified table identifier.
+	TableName string `json:"tableName"`
+	// Table is a bigtable.Table instance with an eventual open connection
+	Table *bigtable.Table `json:"table"`
+	// Type is the encoded schema type.
+	Type beam.EncodedType `json:"type"`
+}
+
+func (f *writeFn) StartBundle() error {
+	var err error
+	f.Client, err = bigtable.NewClient(context.Background(), f.Project, f.InstanceID)
+	if err != nil {
+		return fmt.Errorf("could not create data operations client: %v", err)
+	}
+
+	f.Table = f.Client.Open(f.TableName)
+	return nil
+}
+
+func (f *writeFn) FinishBundle() error {
+	if err := f.Client.Close(); err != nil {
+		return fmt.Errorf("could not close data operations client: %v", err)
+	}
+	return nil
+}
+
+func (f *writeFn) ProcessElement(key int, values func(*Mutation) bool) error {
+
+	var mutation Mutation
+	for values(&mutation) {
+		
+		err := f.Table.Apply(context.Background(), mutation.RowKey, getBigtableMutation(mutation))
+		if err != nil {
+			return fmt.Errorf("could not apply mutation for row key='%s': %v", mutation.RowKey, err)
+		}
+
+	}
+
+	return nil
+}
+
+type writeBatchFn struct {
+	// Project is the project
+	Project string `json:"project"`
+	// InstanceID is the bigtable instanceID
+	InstanceID string `json:"instanceId"`
+	// Client is the bigtable.Client
+	Client *bigtable.Client `json:"client"`
+	// TableName is the qualified table identifier.
+	TableName string `json:"tableName"`
+	// Table is a bigtable.Table instance with an eventual open connection
+	Table *bigtable.Table `json:"table"`
+	// MaxOpsPerMutation indicates how many operations are contained within a bigtableio.Mutation at max
+	MaxOpsPerMutation uint `json:"maxOpsPerMutation"`
+	// Type is the encoded schema type.
+	Type beam.EncodedType `json:"type"`
+}
+
+func (f *writeBatchFn) StartBundle() error {
+	var err error
+	f.Client, err = bigtable.NewClient(context.Background(), f.Project, f.InstanceID)
+	if err != nil {
+		return fmt.Errorf("could not create data operations client: %v", err)
+	}
+
+	f.Table = f.Client.Open(f.TableName)
+	return nil
+}
+
+func (f *writeBatchFn) FinishBundle() error {
+	if err := f.Client.Close(); err != nil {
+		return fmt.Errorf("could not close data operations client: %v", err)
+	}
+	return nil
+}
+
+func (f *writeBatchFn) ProcessElement(key int, values func(*Mutation) bool) error {
+	maxMutationsPerBatch := 100000 / int(f.MaxOpsPerMutation)
+
+	var rowKeys []string
+	var mutations []*bigtable.Mutation
+
+	mutationsAdded := 0
+
+	var mutation Mutation
+	for values(&mutation) {
+
+		rowKeys = append(rowKeys, mutation.RowKey)
+		mutations = append(mutations, getBigtableMutation(mutation))
+		mutationsAdded += int(f.MaxOpsPerMutation)
+
+		if (mutationsAdded + int(f.MaxOpsPerMutation)) > maxMutationsPerBatch {
+			err := tryApplyBulk(f.Table.ApplyBulk(context.Background(), rowKeys, mutations))
```

Comment:
ProcessElement and other bundle Lifecycle methods can have a context.Context parameter for client use, and this would avoid constantly using context.Background(), while allowing beam to add more interesting telemetry in the future.

Reply:
Refactored this bit to have context.Context as parameter of ProcessElement.
Thanks for this hint :))